### PR TITLE
TINKERPOP-1288 & TINKERPOP-1290: VertexProgramInterceptor and GraphComputer TraversalStrategies

### PR DESCRIPTION
**Pull request by okram**
https://issues.apache.org/jira/browse/TINKERPOP-1288
https://issues.apache.org/jira/browse/TINKERPOP-1290

We now have the concept of `TraversalStrategies` for `GraphComputers`. A clear pattern has been established for their use with `SparkGraphComputer`. Extending these concepts to `TinkerGraphComputer` will be easy. Moreover, we have `SparkInterceptorStrategy` which we can continue to extend with traversal patterns that we can map to use Spark DSL under the hood instead of `GraphComputer` semantics for performance boosting. However, after some intelligent refactoring around partitioning and caching (thanks @RussellSpitzer), `SparkGraphComputer` is very close in speed to Spark DSL computations. This is good as if that was sufficiently large, that would destroy the purpose of writing Gremlin traversals for analytics. Either way, we are both fast with non-intercepted traversals and intercepted traversal. In short, this is an epic PR.

CHANGELOG (master)

```
* Added `VertexProgramInterceptor` interface as a general pattern for `GraphComputer` providers to use for bypassing `GraphComputer` semantics where appropriate.
* Added `SparkStarBarrierInterceptor` that uses Spark DSL for local star graph traversals that end with a `ReducingBarrierStep`.
* Added `SparkInterceptorStrategy` which identifies which interceptor to use (if any) given the submitted `VertexProgram`.
* Added `SparkSingleIterationStrategy` that does not partition nor cache the graph RDD if the traversal does not message pass.
* Added more helper methods to `TraversalHelper` for handling scoped traversal children.
```

I'm going to cherry pick the bug fix to `TraversalHelper.isLocalStarGraph()` to tp31/.

CHANGELOG (tp31)

```
* Fixed a severe bug in `TraversalHelper.isLocalStarGraph()`.
```

---

BENCHMARKS (Friendster on the Blades):

```
g.V().count()
  before: 4.5 minutes
  no-partition: 2.9 minutes 
  interceptor: 2.5 minutes

g.V().out().count()
  before: 10 minutes
  no-partitioner: 3.1 minutes 
  interceptor: 2.6 minutes

g.V().groupCount().by(outE().count())
  before: 12 minutes
  no-partitioner: 3.2 minutes
  interceptor: 2.6 minutes 
```

---

VOTE +1 (`mvn clean install` + integration tests -- Giraph still running)

### CODE REVIEW

**Comment by okram** [comment]
```
[INFO] Apache TinkerPop .................................. SUCCESS [4.770s]
[INFO] Apache TinkerPop :: Gremlin Shaded ................ SUCCESS [2.330s]
[INFO] Apache TinkerPop :: Gremlin Core .................. SUCCESS [40.270s]
[INFO] Apache TinkerPop :: Gremlin Test .................. SUCCESS [11.526s]
[INFO] Apache TinkerPop :: Gremlin Groovy ................ SUCCESS [42.358s]
[INFO] Apache TinkerPop :: Gremlin Groovy Test ........... SUCCESS [6.934s]
[INFO] Apache TinkerPop :: TinkerGraph Gremlin ........... SUCCESS [3:31.823s]
[INFO] Apache TinkerPop :: Gremlin Benchmark ............. SUCCESS [5.559s]
[INFO] Apache TinkerPop :: Hadoop Gremlin ................ SUCCESS [10:27.283s]
[INFO] Apache TinkerPop :: Spark Gremlin ................. SUCCESS [6:53.817s]
[INFO] Apache TinkerPop :: Giraph Gremlin ................ SUCCESS [3:00:08.742s]
[INFO] Apache TinkerPop :: Neo4j Gremlin ................. SUCCESS [21:48.184s]
[INFO] Apache TinkerPop :: Gremlin Driver ................ SUCCESS [10.726s]
```

@spmallette -- `GremlinDriverIntegrateTest.shouldGetOneThenSomeThenSomeMore()` failed. Seems nothing to do with `GraphComputer`/Spark. ??

**Comment by spmallette** [comment]
That has nothing to do with spark. That test seems to fail on some systems some of the times (of course, never fails locally so that i can observe it happening). I will look to harden that test again - I've made a note to do so.

**Comment by spmallette** [comment]
Ran `docker/build.sh -t -i` and all of it worked nicely. Didn't even get the gremlin server integration test fail. not sure why travis tanked - @okram  you might want to look at that - probably nothing.

VOTE +1

File: spark-gremlin/src/main/java/org/apache/tinkerpop/gremlin/spark/process/computer/traversal/strategy/optimization/interceptor/SparkStarBarrierInterceptor.java [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+        final ReducingBarrierStep endStep = (ReducingBarrierStep) traversal.getEndStep(); // needed for the final traverser generation
+        traversal.removeStep(0);                                    // remove GraphStep
+        traversal.removeStep(traversal.getSteps().size() - 1);      // remove ReducingBarrierStep
+        traversal.applyStrategies();                                // compile
+        boolean identityTraversal = traversal.getSteps().isEmpty(); // if the traversal is empty, just return the vertex (fast)
+        ///////////////////////////////
+        ((MemoryTraversalSideEffects) traversal.getSideEffects()).setMemory(memory, true); // any intermediate sideEffect steps are backed by SparkMemory
+        memory.setInExecute(true);
+        final JavaRDD<Traverser.Admin<Object>> nextRDD = inputRDD.values()
+                .filter(vertexWritable -> ElementHelper.idExists(vertexWritable.get().id(), graphStepIds)) // ensure vertex ids are in V(x)
```
Comment (twilmes) [inline spark-gremlin/src/main/java/org/apache/tinkerpop/gremlin/spark/process/computer/traversal/strategy/optimization/interceptor/SparkStarBarrierInterceptor.java]: Could this step be conditionally added only if there are ids to filter on?

Comment (okram) [inline spark-gremlin/src/main/java/org/apache/tinkerpop/gremlin/spark/process/computer/traversal/strategy/optimization/interceptor/SparkStarBarrierInterceptor.java]: Originally I had it like that, but the times were "the same" on the Friendster graph (2.5 billion edges). I decided to remove that logic to make the code clear especially as we advance this class moving forward. The `ElementHelper.idExists()` simply returns `true` when there are no ids as it does `if(ids.length == 0) return true`. I suspect the JVM optimizes this all away very quickly.

Comment (okram) [inline spark-gremlin/src/main/java/org/apache/tinkerpop/gremlin/spark/process/computer/traversal/strategy/optimization/interceptor/SparkStarBarrierInterceptor.java]: By "moving this class forward," I mean when we handle `g.E()` ... as you can see, the logic will start to get gross fast with a 4-condition branch. I thought it best to get it all clean and clear so its understandable whats going on.

**Comment by twilmes** [comment]
Excellent stuff 
VOTE: +1

**Pull request merged**